irmago_test.go

```diff
@@ -798,3 +799,228 @@ func TestPrivateKeyRings(t *testing.T) {
 	require.NoError(t, err)
 }
+
+// Helper functions for wizard tests below
+func credid(s string) CredentialTypeIdentifier {
+	return NewCredentialTypeIdentifier(s)
+}
+func credidptr(s string) *CredentialTypeIdentifier {
+	id := credid(s)
+	return &id
+}
+func credinfo(id string) *CredentialInfo {
+	i := credid(id)
+	return &CredentialInfo{
+		SchemeManagerID: i.Root(),
+		IssuerID:        i.IssuerIdentifier().Name(),
+		ID:              i.Name(),
+	}
+}
+func credtype(id string, deps ...string) *CredentialType {
+	i := credid(id)
+	d := CredentialDependencies{{{}}}
+	for _, dep := range deps {
+		d[0][0] = append(d[0][0], credid(dep))
+	}
+	return &CredentialType{
+		SchemeManagerID: i.Root(),
+		IssuerID:        i.IssuerIdentifier().Name(),
+		ID:              i.Name(),
+		Dependencies:    d,
+	}
+}
+func credwizarditem(id string) IssueWizardItem {
+	return IssueWizardItem{Type: IssueWizardItemTypeCredential, Credential: credidptr(id)}
+}
+
+func TestWizardDependencies(t *testing.T) {
+	dependencies := CredentialDependencies{
+		{
+			{credid("a.a.a"), credid("a.a.b")},
+			{credid("a.b.a")},
+		},
+		{
+			{credid("b.a.a")},
+			{credid("b.b.a"), credid("b.b.b")},
+		},
+	}
+
+	wizardcontents := dependencies.WizardContents()
+	require.Equal(t,
+		IssueWizardContents{
+			{
+				{credwizarditem("a.a.a"), credwizarditem("a.a.b")},
+				{credwizarditem("a.b.a")},
+			},
+			{
+				{credwizarditem("b.a.a")},
+				{credwizarditem("b.b.a"), credwizarditem("b.b.b")},
+			},
+		},
+		wizardcontents,
+	)
+
+	conf := &Configuration{CredentialTypes: map[CredentialTypeIdentifier]*CredentialType{}}
+	for _, discon := range dependencies {
+		for _, con := range discon {
+			for _, cred := range con {
+				conf.CredentialTypes[cred] = credtype(cred.String())
+			}
+		}
+	}
+	tests := []struct {
+		creds  map[CredentialTypeIdentifier]struct{}
+		wizard []IssueWizardItem
+	}{
+		{
+			creds: nil,
+			wizard: []IssueWizardItem{
+				credwizarditem("a.a.a"), credwizarditem("a.a.b"), credwizarditem("b.a.a"),
+			},
+		},
+		{
+			creds: map[CredentialTypeIdentifier]struct{}{credid("a.a.a"): {}},
+			wizard: []IssueWizardItem{
+				credwizarditem("a.a.a"), credwizarditem("a.a.b"), credwizarditem("b.a.a"),
+			},
+		},
+		{
+			creds: map[CredentialTypeIdentifier]struct{}{credid("a.b.a"): {}},
+			wizard: []IssueWizardItem{
+				credwizarditem("a.b.a"), credwizarditem("b.a.a"),
+			},
+		},
+		{
+			creds: map[CredentialTypeIdentifier]struct{}{credid("b.b.a"): {}},
+			wizard: []IssueWizardItem{
+				credwizarditem("a.a.a"), credwizarditem("a.a.b"), credwizarditem("b.a.a"),
+			},
+		},
+		{
+			creds: map[CredentialTypeIdentifier]struct{}{credid("b.b.a"): {}, credid("b.b.b"): {}},
+			wizard: []IssueWizardItem{
+				credwizarditem("a.a.a"), credwizarditem("a.a.b"), credwizarditem("b.b.a"), credwizarditem("b.b.b"),
+			},
+		},
+	}
+
+	for _, tst := range tests {
+		require.Equal(t, tst.wizard, wizardcontents.Choose(conf, tst.creds))
+	}
+}
```

Comment:
Suppose you have a condiscon `[[["a.a.a", "a.a.b"], ["a.b.a", "a.b.b", "a.b.c"]]]` and you already have `a.b.b` and `a.b.c` then the path of least resistance would be to choose the `["a.b.a", "a.b.b", "a.b.c"]` part of the disjunction. I think the current code would pick the first one, since both disjunctions do not satisfy and therefore the first one is chosen. Am I right?

Not that always choosing the path of least resistance is in any case logical. If the first conjunction is longer than the second, then always the second would be favoured above the first in that scenario. That is maybe also a bit strange. I was just wondering.

Reply:
Yes, that's right. For now I just pick the first option whenever none of the options is already satisfied, because that is the easiest to implement and the feature already has quite some complexity as it is. We may make it smarter as you suggest in a later phase.